2-collectors/scala-stream-collector/eventhub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/EventHubSink.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2013-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream.sinks
+
+import java.util.concurrent.ScheduledExecutorService
+import java.util.function.BiFunction
+
+import com.microsoft.azure.eventhubs.{ConnectionStringBuilder, EventData, EventHubClient}
+import com.snowplowanalytics.snowplow.collectors.scalastream.model.EventHub
+import scalaz._
+
+/** EventHubSink companion object with factory method */
+object EventHubSink {
+
+  /**
+   * Creates an EventHubSink
+   * Exists so that no threads can get a reference to the EventHubSink during its construction
+   * TODO: rm scalaz \/ once 2.12
+   */
+  def createAndInitialize(
+                           eventHubConfig: EventHub,
+                           streamName: String,
+                           domainName: String,
+                           sasKeyName: String,
+                           sasKey: String,
+                           executorService: ScheduledExecutorService
+  ): \/[Throwable, EventHubSink] = {
+
+    val connStrBuilder = new ConnectionStringBuilder()
+      .setEndpoint( streamName, domainName)
+      .setSasKeyName(sasKeyName)
+      .setSasKey(sasKey)
+
+    val client = \/.fromTryCatch(EventHubClient.createSync(connStrBuilder.toString, executorService))
+
+    client.map{c =>
+      new EventHubSink(c, streamName, executorService)
+    }
+  }
+}
+
+/**
+ * Eventhub Sink for the Scala collector.
+ */
+class EventHubSink private(
+  client: EventHubClient,
+  streamName: String,
+  executorService: ScheduledExecutorService
+) extends Sink {
+
+  // Records must not exceed MaxBytes - 1MB
+  val MaxBytes = 1000000L
+
+  implicit lazy val ec = concurrent.ExecutionContext.fromExecutorService(executorService)
+
+  private val biFun = new BiFunction[Void, Throwable, Void] {
+    override def apply(v: Void, t: Throwable): Void = {
+      if (t != null) sys.error(s"Sending event to EventHub failed: ${t.getMessage}")
+      null
+    }
+  }
```

Comment:
In your opinion, how hard would it be to build some kind of retrying logic in this function?